### PR TITLE
Add user agent override list support. (fixes #488)

### LINKED ISSUE #488: Implement UA string site override list functionality
We will want to implement an override list for 1.1 that allows us to send a different UA string on a per URL basis for compatibility reasons. We may also want to implement the ability to select Servo as a rendering engine on a per URL basis for use in rendering experiments and testing new web platform features.

### PR DESCRIPTION
**Pull request by bluemarvin**

### CODE REVIEW

File: app/src/common/shared/org/mozilla/vrbrowser/browser/UserAgentOverride.java [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+            checkedDomains[ix] = domain;
+            override = mOverrideCache.get(domain);
+            if (override != null) {
+                Log.d(LOGTAG, "found cached override: " + override);
+                addToCache(checkedDomains, override);
+                return override.equals(NO_OVERRIDE_FOUND) ? null : override;
+            }
+            String domainHash = hashDomain(domain);
+            if (domainHash == null) {
+                Log.d(LOGTAG, "Failed to hash domain" + domain);
```
Comment (cvan) [inline app/src/common/shared/org/mozilla/vrbrowser/browser/UserAgentOverride.java]: `: ` in log

File: app/src/common/shared/org/mozilla/vrbrowser/browser/UserAgentOverride.java [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        if (fullDomain == null) {
+            return null;
+        }
+
+        fullDomain = fullDomain.toLowerCase();
+
+        String override = mOverrideCache.get(fullDomain);
+
+        if (override != null) {
+            Log.d(LOGTAG, "Found override:" + override);
```
Comment (cvan) [inline app/src/common/shared/org/mozilla/vrbrowser/browser/UserAgentOverride.java]: space after `:`

File: app/src/common/shared/org/mozilla/vrbrowser/browser/UserAgentOverride.java [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    private String hashDomain(String aDomain) {
+        try {
+            MessageDigest md = MessageDigest.getInstance("SHA-512");
+            byte[] digest = md.digest(aDomain.getBytes());
+            StringBuilder sb = new StringBuilder();
+            for (int value: digest) {
+                sb.append(Integer.toString((value & 0xff) + 0x100, 16).substring(1));
+            }
+
+            return sb.toString().toUpperCase();
```
[the remark this answers is not in the document]
Comment (bluemarvin) [reply app/src/common/shared/org/mozilla/vrbrowser/browser/UserAgentOverride.java]: If you want to send me a file with lower case SHA values, that works for me :)

File: app/src/common/shared/org/mozilla/vrbrowser/browser/UserAgentOverride.java (hunk 1 above)
Comment (bluemarvin) [reply app/src/common/shared/org/mozilla/vrbrowser/browser/UserAgentOverride.java]: Done

File: app/src/common/shared/org/mozilla/vrbrowser/browser/UserAgentOverride.java (hunk 2 above)
Comment (bluemarvin) [reply app/src/common/shared/org/mozilla/vrbrowser/browser/UserAgentOverride.java]: Done

File: app/src/common/shared/org/mozilla/vrbrowser/browser/UserAgentOverride.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (bluemarvin) [reply app/src/common/shared/org/mozilla/vrbrowser/browser/UserAgentOverride.java]: Updated to use lowercase.

**Comment by cvan** [comment]
this looks good to go! excellent work :+1:

**Comment by bluemarvin** [comment]
@cvan, I think this is ready to land. Should I wait until you have time to update userAgentOverride.json? Or should that be a followup?

**Comment by cvan** [comment]
> @cvan, I think this is ready to land. Should I wait until you have time to update userAgentOverride.json? Or should that be a followup?

to make which changes? you mean to add more exceptions?

**Comment by bluemarvin** [comment]
> > @cvan, I think this is ready to land. Should I wait until you have time to update userAgentOverride.json? Or should that be a followup?
> 
> to make which changes? you mean to add more exceptions?

Yeah, I wasn't sure if this was the complete file or not.

**Comment by cvan** [comment]
yep, that's the complete file - thanks for checking

**Pull request merged**